### PR TITLE
Django crud api

### PR DESCRIPTION
**Pull request by paulodhiambo**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      

#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by paulodhiambo** [comment]
Fixes #557

**Comment by hectorkambow** [comment]
👋  Thank you for submitting this PR - we will review it ASAP - all PRs go into our queue will be looked over before being approved to be published.

**Comment by adrianmurage** [comment]
@hectorkambow 
Kindly run a plagiarism check on this

**Comment by adrianmurage** [comment]
Hi @paulodhiambo 
I'll be conducting the review in two stages:
- Preliminary review where I'll look out for: grammar, spelling, and readability
- Tech Review where I'll look out for: technical accuracy, beginner-friendliness, and currency

**Comment by hectorkambow** [comment]
plagiarism check done

File: articles/django-crud-api/index.md [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+     |████████████████████████████████| 957 kB 595 kB/s 
+Requirement already satisfied: django>=2.2 in ./venv/lib/python3.8/site-packages (from djangorestframework) (3.1.3)
+Requirement already satisfied: asgiref<4,>=3.2.10 in ./venv/lib/python3.8/site-packages (from django>=2.2->djangorestframework) (3.3.1)
+Requirement already satisfied: sqlparse>=0.2.2 in ./venv/lib/python3.8/site-packages (from django>=2.2->djangorestframework) (0.4.1)
+Requirement already satisfied: pytz in ./venv/lib/python3.8/site-packages (from django>=2.2->djangorestframework) (2020.4)
+Installing collected packages: djangorestframework
+Successfully installed djangorestframework-3.12.2
+```
+
+Lets create out django-todo project now by running the command `django-admin startproject django_todo`
```
Comment (adrianmurage) [inline articles/django-crud-api/index.md:63]: Inconsistency in the project naming
"django-todo" and "django_todo" pick one and use it consistently

Also correct 
"Let's create our..."

File: articles/django-crud-api/index.md [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    title = models.CharField(max_length = 100)
+    body = models.CharField(max_length = 100)
+    is_completed = models.BooleanField(default=False)
+    date_created = models.DateField(auto_created=True)
+    last_modified = models.DateField(auto_now=True)
+
+    def __str___(self):
+        return self.title 
+```
+To create an SQLite database and Todo table in the database, we will run the command `./manage.py makemigrations` to create SQL queries that will be used to create a `Todo` table in the database. To create the Todo table in the database will run the command `./manage.py migrate`.
```
Comment (adrianmurage) [inline articles/django-crud-api/index.md:124]: The idea around migrations is a little confusing, even for me. Consider pointing the reader to the [migration docs](https://docs.djangoproject.com/en/3.1/topics/migrations/).
Then rewrite these as sequential steps like so:

To make migrations run:
`./manage.py makemigrations`

To apply the migrations run:
`./manage.py migrate`

File: articles/django-crud-api/index.md [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+class TodoSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Todo
+        fields = "__all__"
+```
+
+We only need to pass in the model that we need to serialize and the fields to serialize as meta fields. In our case, we are serializing all of the fields from the `Todo` model that's why we are passing `__all__` to our `fields` variable.
+
+### Creating the TodoAPIView
+In the `todo` app directory in the `views.py` file, we need to write in the logic for the CRUD functionality for our app. Django Rest framework comes with inbuilt classes which make building the CRUD functionality very easy. We start by importing the Django rest framework classes that we will use to create our CRUD API. We just need to create a class for each of our crud endpoints and add in the `queryset` and the `serializer_class` for our `Todo` model.
```
Comment (adrianmurage) [inline articles/django-crud-api/index.md:144]: Consider rearranging this to:
"In the `views.py` file in our `todo` app directory, we are going to write the logic for the CRUD functionality for our app."

**Review by adrianmurage — CHANGES_REQUESTED** [review]
Hi @paulodhiambo, great article. I have left some grammatical and readability changes you should consider making.
I'll go ahead and start reviewing the technical correctness of the article as you make them.

File: articles/django-crud-api/index.md [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+from django.urls import include
+
+urlpatterns = [
+    path('admin/', admin.site.urls),
+    path('api/v1/todo/',include("todo.urls"))
+]
+```
+### Testing the endpoints
+
+Making a GET request to http://localhost:8000/api/v1/todo in postman returns a list of `todos`.
```
Comment (adrianmurage) [inline articles/django-crud-api/index.md:231]: I feel like you should point out that the server should be running.
Consider addding an instruction to run the command that runs the server.

Comment (adrianmurage) [inline articles/django-crud-api/index.md:231]: There was no todo items added by this point in the article.
Making a GET request returns an empty list

Comment (adrianmurage) [reply articles/django-crud-api/index.md:231]: I suggest you re-order the requests:
1. POST
2. GET
3. PUT
4. DELETE

Comment (adrianmurage) [reply articles/django-crud-api/index.md:231]: Also consider adding images with the various operations on Postman to help the reader identify what they need to be doing if they happen to be unfamiliar with postman

**Review by adrianmurage — CHANGES_REQUESTED** [review]
@paulodhiambo 
Great article. I have pointed out a few improvements.
Kindly make them and when you're done ping me, I'll go over the article one last time and mark it ready for final review

**Comment by paulodhiambo** [comment]
@adrianmurage I am done with the suggested changes.

**Review by adrianmurage — APPROVED** [review]
@paulodhiambo Great job with this!
@hectorkambow this article is ready for final review

**Pull request merged**